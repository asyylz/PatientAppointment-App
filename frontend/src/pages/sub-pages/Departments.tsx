import CustomButton from '../../components/UI/CustomButton';
import classes from './Department.module.css';
import { useDispatch, useSelector } from 'react-redux';
import { RootState, AppDispatch } from './../../store/index';
import { fetchDepartments } from './../../store/departments-slice';
import { useEffect } from 'react';

const Departments: React.FC = () => {
  const dispatch: AppDispatch = useDispatch();

  const {
    entities:  departments ,
    status,
    error,
  } = useSelector((state: RootState) => state.departments);


  useEffect(() => {
    if (status === 'idle') {
      dispatch(fetchDepartments());
    }
  }, [status, dispatch]);

  return (
    <div className={classes.mainWrapper} 
    //style={{ border: '1px solid red' }}
    >
      <div
        //style={{border:'3px solid green'}}
        className={classes.headerWrapper}
      >
        <h1
        //style={{border:'1px solid red'}}
        >
          DEPARTMENTS
        </h1>
        <hr />
      </div>

      <div
        className={classes.container}
        //style={{ border: '2px solid red' }}
      >
        {status === 'loading' && <p>Loading...</p>}
        
        {status === 'succeeded' &&
          departments.map((department) => (
            <CustomButton
              key={department._id}
              department={department as Department}
            />
          ))}
        {status === 'failed' && <p>{error}</p>}
      </div>
    </div>
  );
};

export default Departments;
