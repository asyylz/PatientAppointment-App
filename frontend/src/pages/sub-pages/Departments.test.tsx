import renderer from 'react-test-renderer';
import { act } from 'react';
import '@testing-library/jest-dom';
import { render, waitFor, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { combineReducers } from 'redux';
import deparmentsReducer from '../../store/departments-slice';
import { configureStore } from '@reduxjs/toolkit/react';
//import * as router from 'react-router-dom';
import Departments from './Departments';
import searchReducer from '../../store/search-slice';

const mockAllReducers = combineReducers({
  departments: deparmentsReducer,
  search: searchReducer,
  // other reducers...
});

const store = configureStore({
  reducer: mockAllReducers,
});

console.log(store.getState());
describe('Departments Page', () => {
  beforeEach(() => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Departments />
        </MemoryRouter>
      </Provider>
    );
  });
  /* -------------------------- - ------------------------- */
  it('1--Matches the DOM snapshot', () => {
    const tree = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <Departments />
        </MemoryRouter>
      </Provider>
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  /* -------------------------- - ------------------------- */
  it('2--After rendering useEffect fetch departments with status "loading', async () => {
    console.log(store.getState().departments.status);
    expect(store.getState().departments.status).toEqual('loading');

    expect(
      store.dispatch({
        type: 'departments/fetch/pending',
      })
    ).toHaveBeenCalled;
    expect(store.getState().departments.status).toEqual('loading');
    await waitFor(() => {
      expect(screen.getByText('Departments loading...'));
    });
    console.log(store.getState().departments.status);
  });

  /* -------------------------- - ------------------------- */
  it('fetchDepartments action fulfilled and screen expect  related data', async () => {
    await act(async () => {
      store.dispatch({
        type: 'departments/fetch/fulfilled',
        payload: [
          {
            _id: '6660fbd12e3c00fe18cfceec',
            departmentMain: 'Allergy & Immunology',
            departmentSub: [
              'Allergy',
              'Allergy & Immunology',
              'Immunology',
              'Immunopathology',
              'Pediatric Allergy',
            ],
          },
          {
            _id: '6660fbd12e3c00fe18cfceed',
            departmentMain: 'Anesthesiology',
            departmentSub: [
              'Anesthesiology',
              'Critical Care Anesthesiology',
              'Critical Care Medicine',
              'Pain Management',
            ],
          },
        ],
      });
    });
    expect(store.getState().departments.status).toEqual('succeeded');
    expect(screen.getAllByText('Allergy & Immunology')[0]).toBeInTheDocument();
    expect(screen.getAllByText('Anesthesiology')[0]).toBeInTheDocument();
  });
  /* -------------------------- - ------------------------- */
  it('fetchDepartments action rejected and screen expect error message', async () => {
    await act(async () => {
      store.dispatch({
        type: 'departments/fetch/rejected',
        error: new Error('Network Error'),
      });
    });
    expect(store.getState().departments.status).toEqual('failed');
    expect(screen.getByText('Network Error')).toBeInTheDocument();
  });
  /* -------------------------- - ------------------------- */
  it('fetchDepartments action rejected and screen expect error message comes from payload', async () => {
    await act(async () => {
      store.dispatch({
        type: 'departments/fetch/rejected',
        payload: {
          message: 'Payload error',
        },
      });
    });
    expect(store.getState().departments.status).toEqual('failed');
    expect(screen.getByText('Payload error')).toBeInTheDocument();
  });
});
