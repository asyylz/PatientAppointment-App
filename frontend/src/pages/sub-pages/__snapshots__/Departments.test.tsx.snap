// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Departments Page 1--Matches the DOM snapshot 1`] = `
<div
  className="mainWrapper"
>
  <div
    className="headerWrapper"
  >
    <h2>
      DEPARTMENTS
    </h2>
    <hr />
  </div>
  <div
    className="container"
  >
    <p>
      Departments loading...
    </p>
  </div>
</div>
`;
